### PR TITLE
DirectILLAutoProcess

### LINKED ISSUE #28161: Direct ILL Auto
Turn the reduce.py script for direct geometry TOF instruments to an algorithm: DirectILLAutoProcess

Must support PSD based instruments IN5, PANTHER, SHARP as well as the old IN4 and IN6.

Support 3 process types: Calibration (vanadium or sample at low T), Container and Sample.

### PR DESCRIPTION
**Pull request by darominski**
This PR implements new algorithm to automatize the data reduction for direct geometry TOF instruments at the ILL: PANTHER, SHARP, IN5, and retroactively also for IN4 and IN6. 

The algorithm connects already existing DirectILL algorithms: `DirectILLCollectData`, `DirectILLReduction`, `DirectILLIntegrateVanadium`, `DirectILLDiagnostics` in a way that frees user from connecting the inputs and outputs of the each individual step, as well as provide robust options for masking data and full attenuation treatment. The goal is to significantly shorten and simplify the scripts used by the non-expert user, who runs a standarized reduction workflow. For custom workflows, the individual calls to DirectILL family of algorithms in a script may still be needed.

**To test:**

1. Code review.
2. Tests should be passing.
3. Run a minimal full powder reduction of PANTHER data using a script:

```
EmptyRuns = '9777'
VanaRuns = "9406"
SampleRuns = '9738, 9740, 9744'

EmptyName = 'MTCell19meV'
VanaName = "V19meV"
SampleName = 'He3C60'

Ei = 19
Elc = 104

geometry = {
            'Shape': 'FlatPlate',
            'Height': 4.0,
            'Width': 2.0,
            'Thick': 0.2,
            'Angle': 0.0,
            'Center': [0.0, 0.0, 0.0]
}

material = {
            'ChemicalFormula': 'C60-(He3)0.80',
            'ZParameter': 4.0,
            'UnitCellVolume': 2744.0,  # =14A**3 for one Cell CFC C60
            'SamplePackingFraction': 0.1
        }

DirectILLAutoProcess(
            Runs=EmptyRuns,
            OutputWorkspace=EmptyName,
            ProcessAs='Empty',
            ReductionType='Powder',
            IncidentEnergyCalibration="Energy Calibration ON",
            IncidentEnergy=Ei,
            ElasticChannelIndex=Elc,
            EPPCreationMethod='Calculate EPP',
            SaveOutput=False,
            ClearCache=True
)

# vanadium
DirectILLAutoProcess(
            Runs=VanaRuns,
            OutputWorkspace=VanaName,
            ProcessAs='Vanadium',
            ReductionType='Powder',
            EmptyContainerWorkspace=EmptyName,
            FlatBkg='Flat Bkg ON',
            IncidentEnergyCalibration="Energy Calibration ON",
            IncidentEnergy=Ei,
            ElasticChannelIndex=Elc,
            EPPCreationMethod='Calculate EPP',
            SaveOutput=False,
            ClearCache=True
)

# sample
DirectILLAutoProcess(
            Runs=SampleRuns,
            OutputWorkspace=SampleName,
            ProcessAs='Sample',
            ReductionType='Powder',
            VanadiumWorkspace=VanaName,
            FlatBkg='Flat Bkg OFF',
            EmptyContainerWorkspace=EmptyName,
            EPPCreationMethod='Calculate EPP',
            ElasticChannelIndex=Elc,
            MaskedAngles=[0, 20],
            IncidentEnergy=Ei,
            IncidentEnergyCalibration="Energy Calibration ON",
            AbsorptionCorrection='Fast',
            SelfAttenuationMethod='MonteCarlo',
            SampleMaterial=material,
            SampleGeometry=geometry,
            SaveOutput=False,
            ClearCache=True
)

```

You can change parameters for calibrations, and masking options from the provided parameters. There should be no issues.
4. Reduce the data from the example above in DrILL by filling out the table, using provided rundex.
5. Single crystal example:

```
ei = 19.03
elp = 104
Load(Filename='Vnorm012315.nxs', OutputWorkspace='Norm')
GroupWorkspaces(InputWorkspaces='Norm', OutputWorkspace='NormVana')
eBins = [-2., 0.19, 16]  # inelastic
group_by = 4

output_name = 'panther_236'
DirectILLAutoProcess(
            Runs='13806, 13807',
            OutputWorkspace=output_name,
            ProcessAs='Sample',
            ReductionType='SingleCrystal',
            VanadiumWorkspace='NormVana',
            MaskWorkspace='Mask212.xml',
            IncidentEnergyCalibration="Energy Calibration ON",
            IncidentEnergy=ei,
            ElasticChannelIndex=elp,
            EnergyRebinningParams=eBins,
            FlatBkg='Flat Bkg ON',
            GroupDetBy=group_by,
            ClearCache=True,
            SaveOutput=False
)
```
Fixes #28161 .

Release notes and documentation updated. 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by gvardany — COMMENTED** [review]
Just a few comments on the code and tests.
I run the provided example for powder.
It run through and produced workspaces as in the attachment.
Screenshot 2022-02-07 at 13 18 06

There are a few issues with this.
First and most important, is that the SofQW workspaces are not correct. 
They have q as x-axis, but not deltaE as y-axis.
Are you sure those are the right outputs as produced by `DirectILLReduction`?
Similarly, the y-axis of SofTW is not 2 theta in degrees, but just spectrum number.

Then, for vanadium, the naming convention looks a bit strange.
The run numbers are in a different place as compared to the sample outputs.
Then, there is SofQ, which I don't understand. Is this really supposed to be the diffraction pattern, i.e. integrated over all deltaE bins?

There is a dangling workspace, called `corrected` which doesn't belong to any group. If clear cache is true by default, I wasn't expecting it.

Then, if self attenuation corrections are requested (no matter the method), I think we should output also the corrections (the workspace group containing up to 4 terms), even with the clear cache enabled. Those are long to compute, and useful to look at for diagnostics.

Also, I'd suggest adding the units to the observables in the workspace names, just to be more informative and also consistent with other techniques, such as SANS.
For example `.....Ei19mev_T20K` looks much better than now.

Finally, could you provide an example with the single crystal as well, containing at least 2 sample angles?

File: Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/DirectILLAutoProcessTest.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+            SaveOutput=False,
+            ClearCache=True
+        )
+        self.assertTrue(isinstance(mtd[vanadium_name], WorkspaceGroup))
+        self._check_output(mtd[vanadium_name][0], 665, 226, True, "q", "MomentumTransfer", "Energy transfer",
+                           "DeltaE")  # S(Q, w)
+        self._check_output(mtd[vanadium_name][1], 226, 669, True, "Energy transfer", "DeltaE", "Scattering angle",
+                           "Degrees")  # S(2theta, w)
+
+    def test_sample_single_crystal(self):
```
[the remark this answers is not in the document]
Comment (darominski) [reply Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/DirectILLAutoProcessTest.py:69]: Absolutely, I can. I put the sample powder reduction only in the system test due to its length. The unit test takes currently about 15 s, with the powder test included it doubles.  If you think that's fine, I am going to include it.

File: Testing/SystemTests/tests/framework/DirectILLAutoProcessTest.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        ei = 19.03
+        elp = 104
+        Load(Filename='Vnorm012315.nxs', OutputWorkspace='Norm')
+        GroupWorkspaces(InputWorkspaces='Norm', OutputWorkspace='NormVana')
+        eBins = [-2., 0.19, 16]  # inelastic
+        group_by = 4
+
+        output_name = 'panther_236'
+        DirectILLAutoProcess(
+            Runs='13806',
```
[the remark this answers is not in the document]
Comment (darominski) [reply Testing/SystemTests/tests/framework/DirectILLAutoProcessTest.py:156]: Done. Also, example above also includes reduction of single crystal PANTHER data with two different sample positions.

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/DirectILLCollectData.py [hunk 3]
```diff
@@ -738,6 +786,58 @@ def _flatBkgMon(self, monWS):
         self._cleanup.cleanup(monWS)
         return bkgSubtractedMonWS
 
+    def _get_grouping_pattern(self, ws):
```
[the remark this answers is not in the document]
Comment (darominski) [reply Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/DirectILLCollectData.py]: It had only slight differences, and could have been generalized into a method in `DirectILL_common`, which I have just done. Also, I added unit tests for both intra-tube and between-tubes grouping pattern creation for IN5, PANTHER, and SHARP.

**Comment by darominski** [comment]
> Just a few comments on the code and tests. I run the provided example for powder. It run through and produced workspaces as in the attachment. Screenshot 2022-02-07 at 13 18 06
> 
> There are a few issues with this. First and most important, is that the SofQW workspaces are not correct. They have q as x-axis, but not deltaE as y-axis. Are you sure those are the right outputs as produced by `DirectILLReduction`? Similarly, the y-axis of SofTW is not 2 theta in degrees, but just spectrum number.
> 
> Then, for vanadium, the naming convention looks a bit strange. The run numbers are in a different place as compared to the sample outputs. Then, there is SofQ, which I don't understand. Is this really supposed to be the diffraction pattern, i.e. integrated over all deltaE bins?
> 
> There is a dangling workspace, called `corrected` which doesn't belong to any group. If clear cache is true by default, I wasn't expecting it.
> 
> Then, if self attenuation corrections are requested (no matter the method), I think we should output also the corrections (the workspace group containing up to 4 terms), even with the clear cache enabled. Those are long to compute, and useful to look at for diagnostics.
> 
> Also, I'd suggest adding the units to the observables in the workspace names, just to be more informative and also consistent with other techniques, such as SANS. For example `.....Ei19mev_T20K` looks much better than now.
> 
> Finally, could you provide an example with the single crystal as well, containing at least 2 sample angles?

Thank you for the review.

I corrected the naming scheme, and it is now both consistent for sample, empty, and vanadium, and naming also contains the relevant unit for parameters used (K, meV).

The issue with S (Q, w) was found in the grouping. Now, the grouping is performed only for the empty and single crystal sample, which is the workflow this feature was requested. There is no workflow where grouping would be of great use in the powder case, apart from increasing statistics of the flat background estimation, which is a separate thing. In any case, this issue has been fixed.

The dangling 'corrected' actually comes from `PaalmanPingsMonteCarloAbsorption`, but it is properly cleaned up now. Also, the absorption correction workspaces are kept, even if `ClearCache` is checked, as you suggested.

I updated the example of this PR to include the single crystal reduction, and expanded the system test to include two numors.

Other comments that were dispersed through the code have also been answered now. Could have a second look at this?

File: Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/DirectILLAutoProcessTest.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (gvardany) [reply Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/DirectILLAutoProcessTest.py:69]: I think this is critical enough to pay the price. 30 second won't make it the slowest test.

[the remark this answers is not in the document]
Comment (darominski) [reply Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/DirectILLAutoProcessTest.py:69]: It's added then.

File: docs/source/techniques/DirectILL.rst [hunk 4]
```diff
@@ -705,4 +743,101 @@ Output:
...
+And the same reduction, but with the use of the :ref:`algm-DirectILLAutoProcess`:
+
+.. testsetup:: FullExampleAutoProcess
+
+    config.setFacility('ILL')
+    config.appendDataSearchSubDir('ILL/IN4/')
+
+.. testcode:: FullExampleAutoProcess
+
+    vanadium_runs = 'ILL/IN4/085801-085802'
```
Comment (gvardany) [inline docs/source/techniques/DirectILL.rst:755]: Just a nitpick, if the `ILL/IN4` is added as a subdirectory, can we omit them in these paths to make it more appealing to the eye? Also `.nxs` extensions can be skipped.

**Comment by gvardany** [comment]
Thanks for updates, it is much better.

> The issue with S (Q, w) was found in the grouping. Now, the grouping is performed only for the empty and single crystal sample, which is the workflow this feature was requested. There is no workflow where grouping would be of great use in the powder case, apart from increasing statistics of the flat background estimation, which is a separate thing. In any case, this issue has been fixed.

The axis of SQW is fixed now, however, I am not sure I understand the explanation. Detector grouping is irrelevant/plain wrong to run in Q space, as there is no longer per-pixel information available. In case you need to group data in reduced space, you should resort to Rebin, Rebin2D, etc. Can you confirm that in no scenario (powder, crystal, empty, vana, sample) there is no detector grouping attempted on reduced (i.e. Q converted) data?

I also noticed that in the provided example for powder, the attenuation correction workspace group is not present. However, it contains only one workspace. Is this because container geometry and material were not provided (even though container runs were)? Could you perhaps add container geometry and material in the respective test, and make sure that the attenuation correction workspace group contains all the 4 terms in PP formalism?

**Comment by darominski** [comment]
> Thanks for updates, it is much better.
> 
> > The issue with S (Q, w) was found in the grouping. Now, the grouping is performed only for the empty and single crystal sample, which is the workflow this feature was requested. There is no workflow where grouping would be of great use in the powder case, apart from increasing statistics of the flat background estimation, which is a separate thing. In any case, this issue has been fixed.
> 
> The axis of SQW is fixed now, however, I am not sure I understand the explanation. Detector grouping is irrelevant/plain wrong to run in Q space, as there is no longer per-pixel information available. In case you need to group data in reduced space, you should resort to Rebin, Rebin2D, etc. Can you confirm that in no scenario (powder, crystal, empty, vana, sample) there is a detector grouping attempted on reduced (i.e. Q converted) data?
> 
> I also noticed that in the provided example for powder, the attenuation correction workspace group is now present. However, it contains only one workspace. Is this because container geometry and material were not provided (even though container runs were)? Could you perhaps add container geometry and material in the respective test, and make sure that the attenuation correction workspace group contains all the 4 terms in PP formalism?

Completely agreed that grouping after conversion to Q space makes no sense, and is not attempted. However, there would be an issue in `DetectorEfficiencyCorUser` if grouping had been done before `DirectILLReduction` was called, due to mentioned confusion of component names and fetching formula for the efficiency calculation. This runs deeper that this PR and should be thus fixed separately, I shall try to make an issue with a minimal example.

Yes, in the examples I had in hand there was no geometry provided for the container. Checking this out lead me to uncover issues in the `PaalmanPingsMonteCarloAbsorption`, where creating beam profile seem to always fail when sparse instrument is used, and effects of using `SetBeam` to set a beam geometry also seem to be always ignored by this algorithm. This definitely requires further investigation, but I would say it goes beyond this PR.

**Comment by gvardany** [comment]
Ok, then please create the respective issues, to be handled quickly. For the former, it looks like you might need to move the grouping inside `DirectILLReduction`, in any case it must be before SofQW. I still don't what DetectorEfficiencyCorUser has to do with the grouping. In any case, I agree that this PR can advance as is.

**Review by gvardany — DISMISSED** [review]

**Review by gvardany — APPROVED** [review]

**Pull request merged**